src/control/QueuedTrajectoryExecutor.cpp

```diff
@@ -0,0 +1,91 @@
+#include <chrono>
+#include <aikido/control/QueuedTrajectoryExecutor.hpp>
+
+namespace aikido {
+namespace control {
+
+//==============================================================================
+QueuedTrajectoryExecutor::QueuedTrajectoryExecutor(
+    std::unique_ptr<TrajectoryExecutor> executor)
+  : mExecutor{std::move(executor)}, mInProgress{false}, mMutex{}
+{
+  if (!mExecutor)
+    throw std::invalid_argument("Executor is null.");
+}
+
+//==============================================================================
+QueuedTrajectoryExecutor::~QueuedTrajectoryExecutor()
+{
+  // Do nothing.
+}
+
+//==============================================================================
+void QueuedTrajectoryExecutor::validate(trajectory::TrajectoryPtr traj)
+{
+  mExecutor->validate(traj);
+}
+
+//==============================================================================
+std::future<void> QueuedTrajectoryExecutor::execute(
+    trajectory::TrajectoryPtr traj)
+{
+  validate(traj);
+
+  {
+    std::lock_guard<std::mutex> lock(mMutex);
+    DART_UNUSED(lock); // Suppress unused variable warning
+
+    // Queue the trajectory and promise that will need to be set
+    mTrajectoryQueue.push(std::move(traj));
+    mPromiseQueue.emplace(new std::promise<void>());
+    return mPromiseQueue.back()->get_future();
+  }
+}
+
+//==============================================================================
+void QueuedTrajectoryExecutor::step()
+{
+  mExecutor->step();
+
+  std::lock_guard<std::mutex> lock(mMutex);
+  DART_UNUSED(lock); // Suppress unused variable warning
+
+  // If a trajectory was executing, check if it has finished
+  if (mInProgress)
+  {
+    // Return if the trajectory is still executing
+    if (mFuture.wait_for(std::chrono::seconds(0)) != std::future_status::ready)
+      return;
+
+    // The promise corresponding to the trajectory that just finished must be
+    // at the front of its queue.
+    auto promise = mPromiseQueue.front();
+    mPromiseQueue.pop();
+
+    // Propagate the future's value or exception to the caller
+    try
+    {
+      mFuture.get();
+      promise->set_value();
+    }
+    catch (const std::exception& e)
+    {
+      promise->set_exception(std::current_exception());
+    }
```

Comment:
Why manually copy the result of `mFuture` to `promise`, rather than directly returning `mFuture` to the caller?

Reply:
I think there are two types of futures: ones from a promise made by the `QueuedTrajectoryExecutor` to the client, and ones from a promise made by the underlying `TrajectoryExecutor` to the `QueuedTrajectoryExecutor`.

I believe this distinction is necessary because the first type must be returned immediately after the client calls `QueuedTrajectoryExecutor::execute` so that the client has something to `wait` on. However, the second type isn't created until the underlying `TrajectoryExecutor::execute` is called (when the trajectory is popped off the queue), so it can't be returned immediately by `QueuedTrajectoryExecutor::execute`.

I think this means that I have to manually copy the result of `mFuture` to `promise`, but if there's another way I'd be happy to try that too!

In this implementation, `mPromiseQueue` holds all the promises that have been made by the `QueuedTrajectoryExecutor` to the client, while `mFuture` is the current future that the `QueuedTrajectoryExecutor` is waiting on (from the underlying `TrajectoryExecutor`).